Fixes #638 Implement support for get/setNetworkTimeout().

brettwooldridge opened the pull request:
HikariCP logs a warning at connection pool startup when a driver is found to not support get/setNetworkTimeout() -- due to vulnerability to getting stuck during a network disruption or partition event.

Given our user-base of several hundred thousand users, the warning message has turned into a support burden, with users concerned about its meaning.  One solution would be to do away with the warning, but that doesn't really serve the user.

So, here is the implementation of said feature -- I should have done this a year ago out of self-preservation.

brettwooldridge commented:
@vlsi I am not sure why this was originally flagged as "concurrency hard".  Ultimately, this call just boils down to a ``Socket.setSoTimeout()`` call.  Setting the timeout affects the next blocking operation.  

The primary purpose of the addtion to the JDBC specification was to prevent connections hung by unacknowledged TCP packets due to a network disruption.  It is no substitue for a query timeout, and it is ultimately easy for a user to shoot themselves in the foot.  Fortunately, most users rarely if ever call this directly (clearly pgjdbc users have *never* called it), but infrastructure libraries like HikariCP make use of it in a controlled manner.

vlsi commented:
@brettwooldridge , the question is what would you do in case of TimeoutException?
Do you expect the connection to automatically terminate?

brettwooldridge commented:
@vlsi Ah, that is actually a good question.  I assumed the existing code paths would already handle socket-level errors.  According to the JavaDoc:
> Sets the maximum period a Connection or objects created from the Connection will wait for the database to reply to any one request. If any request remains unanswered, the waiting method will return with a SQLException, and the Connection or objects created from the Connection will be marked as closed. Any subsequent use of the objects, with the exception of the close, isClosed or Connection.isValid methods, will result in a SQLException.

If you want, I can investigate the various code paths to ensure that the exception is handled properly.

brettwooldridge commented:
@vlsi Well ouch, that was not a very happy string to pull. 😭 

I thought I would throw in [some unit tests](https://github.com/brettwooldridge/pgjdbc/blob/master/pgjdbc/src/test/java/org/postgresql/test/jdbc4/NetworkTimeoutTest.java).  But this one is failing:
```java
try {
   _conn.setNetworkTimeout(null, (int) TimeUnit.SECONDS.toMillis(1));
   stmt = _conn.createStatement();
   stmt.execute("SELECT pg_sleep(2)");
   fail("Connection.setNetworkTimeout() did not throw expected exception");
}
catch (SQLException e) {
   assertTrue(stmt.isClosed());       <--- HERE
   assertTrue(_conn.isClosed());
}
```
If that line is removed, the test passes.

My assumption was that the ``Connection`` would be closed by a socket-level exception -- and it *is*, which is Good<sup>tm</sup>.

So, then I added the assertion that the ``Statement`` had been closed as well.  It is *not*.  This is Bad<sup>tm</sup>...

According to the [JDBC Specification](http://java.cnam.fr/iagl/biblio/spec/jdbc-3_0-fr-spec.pdf):
> All Statement objects will be closed when the connection that created them is closed. However, it is good coding practice for applications to close statements as soon as they have finished processing them.

So looking at [PgConnection:1190](https://github.com/brettwooldridge/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1193) I was expecting to some kind of *Statement tracking*; I have to do [something similar in HikariCP](https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java#L270).  However, it appears that none of the ``create/prepareStatement()`` calls track the ``Statement`` objects in any way -- they create and forget.

The lack of resource tracking precludes honoring the JDBC specification contracts, not only with regard to ``Connections`` closing ``Statements``, but also ``Statements`` closing ``ResultSets``.  A ``Connection.close()`` should cascade down, closing open statements, which turn close open result sets.  Again, from the specification:
> Closing a Statement object will close and invalidate any instances of ResultSet produced by that Statement object.
> ...
> These comments about closing Statement objects apply to PreparedStatement and CallableStatement objects as well.

What I propose is this:
 * Remove the assertion relating to the ``Statement`` state (unclosed), it is unrelated to this pull request.
 * Open an issue to track the um...lack of tracking of Connection managed resources.

kscaldef commented:
This PR seems to be in limbo.  Are there any plans to push forward with it?

davecramer commented:
This is sort of up to @brettwooldridge ?

kscaldef commented:
@davecramer Should I interpret that to mean that the project owners approve of this change and will merge it once:

1. it's been rebased
2. the indentation is fixed to make checkstyle happy
3. the tests are migrated to the jdbc41 test suite, since they are using

davecramer commented:
I'm going to say we won't consider it without it. That said I see no reason to not accept it. Although there was an outstanding question of how to handle the timeout exception.

brettwooldridge commented:
@davecramer @kscaldef I have rebased and moved the NetworkTimeoutTest to the JDBC 4.1 suite.  

The issue of proper resource cleanup is independent of this pull request and is being tracked separately by #855.  I *do* have a branch with changes that address that issue, but it is currently unfinished.

kscaldef commented:
@brettwooldridge can you also adjust the indentation so that checkstyle is happy?  I'm guessing the other failure was a heisentest, but I'm not sure.  Sorry to pester you about this, but this issue (failure to recover from a network outage) has bitten us multiple times, and I'd really love to see it fixed.

brettwooldridge commented:
@kscaldef Sure, I'll take a look this week; 5am here in Tokyo now.

jorsol commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java:
```diff
@@ -1480,12 +1480,31 @@ public void abort(Executor executor) throws SQLException {
 
-  public void setNetworkTimeout(Executor executor, int milliseconds) throws SQLException {
-    throw org.postgresql.Driver.notImplemented(this.getClass(), "setNetworkTimeout(Executor, int)");
+  public void setNetworkTimeout(Executor executor /*not used*/, int milliseconds) throws SQLException {
```
This method should check to see that there is an SQLPermission object before allowing the method to proceed. If a SecurityManager exists and its checkPermission method denies calling setNetworkTimeout, this method throws a java.lang.SecurityException.

brettwooldridge commented:
I think such a check is pointless.  

SecurityException is a perfectly reasonable exception to be thrown; better in fact than a SQLEception that wraps such an exception. Checked exceptions by their nature must be wrapped, but that is not true of unchecked exceptions.

What should definitely *not* occur is to return from this method without setting the timeout and without throwing an exception.

jorsol commented:
@brettwooldridge , well it's probably pointless, but the [jdbc spec/javadoc](https://docs.oracle.com/javase/9/docs/api/java/sql/Connection.html#setNetworkTimeout-java.util.concurrent.Executor-int-) requires it:

> This method checks to see that there is an SQLPermission object before allowing the method to proceed. If a SecurityManager exists and its checkPermission method denies calling setNetworkTimeout, this method throws a java.lang.SecurityException.

A simple check like this should do the trick:
```java
SecurityManager secMgr = System.getSecurityManager();
if (secMgr != null) {
  secMgr.checkPermission(new SQLPermission("setNetworkTimeout"));
}
```

davecramer commented:
Dave Cramer

On 1 October 2017 at 20:20, Jorge Solorzano <notifications@github.com>
wrote:

> @brettwooldridge <https://github.com/brettwooldridge> , well it's
> probably pointless, but the jdbc spec/javadoc
> <https://docs.oracle.com/javase/9/docs/api/java/sql/Connection.html#setNetworkTimeout-java.util.concurrent.Executor-int->
> requires it:
>
> This method checks to see that there is an SQLPermission object before
> allowing the method to proceed. If a SecurityManager exists and its
> checkPermission method denies calling setNetworkTimeout, this method throws
> a java.lang.SecurityException.
>
> A simple check like this should do the trick:
>
> SecurityManager secMgr = System.getSecurityManager();if (secMgr != null) {
>   secMgr.checkPermission(new SQLPermission("setNetworkTimeout"));
> }
>
>
won't it throw one anyway if you don't have permission ?

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pgjdbc/pgjdbc/pull/849#issuecomment-333417324>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAYz9mI_noYU6XZ9AiMquM-nCw4lwbBkks5soCxJgaJpZM4OEWiD>
> .
>

jorsol commented:
> won't it throw one anyway if you don't have permission ?

No, the driver has to explicitly check for it.

davecramer commented:
On 1 October 2017 at 20:29, Jorge Solorzano <notifications@github.com>
wrote:

> won't it throw one anyway if you don't have permission ?
>
> No, the driver has to explicitly check for it.
>

I'm confused. If no exception will be thrown if I don't check for it, why
check for it?

Dave Cramer

jorsol commented:
@davecramer , I'm not aware of the internals of the [Security Manager](https://docs.oracle.com/javase/9/docs/api/java/lang/SecurityManager.html), and in most applications a Security Manager policy is not used or enabled, but the JDBC spec requires that some methods do this check.

There is a [SQLPermission ](https://docs.oracle.com/javase/9/docs/api/java/sql/SQLPermission.html) class that contains more info.

brettwooldridge commented:
@jorsol @davecramer There *is no need* to check this permission.  The SecurityManager *will* throw a ``SecurityException`` whether we check for it or not; that is the point of having a ``SecurityManager``.

The *only* reason we would check if such a permission exists is in the case where the code would somehow need to behave differently upon denial.  That is not the case with ``setNetworkTimeout()`` -- we have no fallback logic -- and therefore such a check is unnecessary.  Trust me, I *do* have code [in another project](https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/HikariConfig.java#L1030) whose behavior is altered in the presence of a SecurityManager, but again that is not the case here.

brettwooldridge commented:
Checkstyle issues fixed for ``NetworkTimeoutTest.java``.

jorsol commented:
@brettwooldridge, I may have misunderstood how the SecurityManager works, but AFAIK this is an implementation code, not a consumer code, so in your case your fallback logic is as consumer code.

I have checked some others drivers and they implement the check, and even the DriverManager wich is a class of the JDK has it.
See http://hg.openjdk.java.net/jdk9/jdk9/jdk/file/a9a5d14bfbb4/src/java.sql/share/classes/java/sql/DriverManager.java#l385

As I said, I have not yet tested it with a SecurityManager enabled, but I just want to be sure that the code works in compliance with what the JDBC spec says, so an additional check that might not be needed is better than an missing check that is needed.

brettwooldridge commented:
@jorsol Ok, I'll buy that.  Check added, consummate with [checks done by other drivers](https://github.com/mysql/mysql-connector-j/blob/b3cda4f864902ffdde495b9df93937c3e20009be/src/com/mysql/jdbc/ConnectionImpl.java#L5499).

brettwooldridge commented:
@davecramer @kscaldef I believe this pull request is ready to go.

jorsol commented:
Is there any reason to hold this on? Is there a tentative milestone to merge this?

johnou commented:
ping @vlsi any changes required to get this pulled in?

DominicGunn commented:
+1 one for this, caused some concern in my team.

davecramer commented:
Have you tested this PR ?

Dave Cramer

On 10 November 2017 at 17:29, Dominic Gunn <notifications@github.com> wrote:

> +1 one for this, caused some concern in my team.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pgjdbc/pgjdbc/pull/849#issuecomment-343604182>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAYz9ihCG13jzXgC_kO8HIDwYNadKhboks5s1M5XgaJpZM4OEWiD>
> .
>

johnou commented:
@davecramer I believe the integration test covers it well enough.

brettwooldridge commented:
🎉
